### DISCUSSION TITLE
Sandbox and Production Environment Handling

### DISCUSSION THREAD

**Discussion by artesianst** [microsoft/PowerPlatformConnectors#2098] [Building Connectors]
Hello All,

What is the best practice for creating a connector for a product that has separate Sandbox and Production environments? Should we create two versions of the connector with Sandbox and Production separation or is there a way to handle this in someway? These two environments have separate auth endpoints.

We tried to search but could not find/understand about the best practice. Thanks for your help.

Amit.

**Comment by troystaylor** [comment]
Hi Amit,
Who are going to be the users of your connector? Is it going to be just a custom connector or are you hoping to have the connector certified? For an Independent Publisher connector, only production environments are allowed. For the other connector types, you can use the Set Host policy with a connection parameter, as shown in the second example on this page: [https://learn.microsoft.com/en-us/connectors/custom-connectors/policy-templates/dynamichosturl/dynamichosturl](https://learn.microsoft.com/en-us/connectors/custom-connectors/policy-templates/dynamichosturl/dynamichosturl)

Url Template | Resulting host url
https://@connectionParameters('cat').abc.xyz | https://bar.abc.xyz

You can use paconn to download the connector, add the connection parameter to apiProperties.json, and then upload the connector back to your environment to test. Then, when your users are creating the connection, then can set which environment to connect to.
If, on the other hand, you want your users to be able to set the environment on each action, you could then use the same policy type, but instead of a connection parameter, you could use a header or query parameter.

Troy

**Reply by artesianst** [reply]
Thanks for your quick response @troystaylor. I'll review and study your suggestions and come back with my questions.

**Reply by Ajinder31** [reply]
Hi @troystaylor, Thank you for your quick response. 

We are aware of Host URL policy, but I think that will be helpful in Action endpoints but won't be helpful in case of dynamic OAuth 2.0 authorize endpoints. Please correct me if I'm wrong. 

We're working on custom connector for a service having different sandbox and production authorization endpoint. We came across Salesforce connector in Power Automate asking user inputs if they want to authenticate against sandbox or production endpoint (Screenshot below).

We're planning to add same functionality in our custom connector instead of creating different connector for sandbox and production authorize endpoint. I wasn't able to figure out how to do it or any documentation related to that.

Can you please help us out and point us to reference documentation to achieve the desired output? 

<img width="421" alt="image" src="https://user-images.githubusercontent.com/52828118/199210194-0a31e12f-3be1-4a98-b563-a41fcda22099.png">

Ajinder
